src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -59,69 +91,52 @@ private bool GetNext()
+                    {
+                        IEnumerator<TSource> enumerator = enumerable.GetEnumerator();
+                        ++state;
+                        if (enumerator.MoveNext())
                         {
-                            _enumerator.Dispose();
-                            _enumerator = _second.GetEnumerator();
-                            if (!_enumerator.MoveNext())
-                            {
-                                Dispose();
-                                return false;
-                            }
-                            state = 3;
-                        }
-                        TSource element = _enumerator.Current;
-                        _set = new Set<TSource>(_comparer);
-                        _set.Add(element);
-                        current = element;
-                        return true;
-                    case 2:
-                        if (GetNext())
+                            SetEnumerator(enumerator);
+                            StoreFirst();
                             return true;
-                        _enumerator.Dispose();
-                        _enumerator = _second.GetEnumerator();
-                        state = 3;
-                        goto case 3;
-                    case 3:
+                        }
+                    }
+                }
+                else if (state > 0)
+                {
+                    for (;;)
+                    {
                         if (GetNext())
                             return true;
-                        break;
+                        IEnumerable<TSource> enumerable = GetEnumerable(state - 1);
+                        if (enumerable == null)
+                            break;
+                        SetEnumerator(enumerable.GetEnumerator());
+                        ++state;
+                    }
                 }
 
                 Dispose();
                 return false;
             }
 
-            public override void Dispose()
-            {
-                if (_enumerator != null)
-                {
-                    _enumerator.Dispose();
-                    _enumerator = null;
-                    _set = null;
-                }
-
-                base.Dispose();
-            }
-
             private Set<TSource> FillSet()
             {
                 Set<TSource> set = new Set<TSource>(_comparer);
-                foreach (TSource element in _first)
-                    set.Add(element);
-                foreach (TSource element in _second)
-                    set.Add(element);
-                return set;
+                for (int index = 0; ; ++index)
+                {
+                    IEnumerable<TSource> enumerable = GetEnumerable(index);
+                    if (enumerable == null)
+                        return set;
+
+                    foreach (TSource item in enumerable)
```

Comment:
We could from one part of what I described as "the absolutely massive flaw in this entire PR that I just noticed" a change ago. If there's a path that still lets that happen, I can't see it, since all construction goes through the check on equivalent equality comparers (either one equality comparer considers the other equal, or they are both the same, or both null, or one null and one Default which null gets turned into).